File: go/testutils/api_handler.go
package testutils

import (
	"net/http"

	types_app "github.com/mabels/runified/types/app"
	"github.com/rs/zerolog"
)

type MockApiHandler[Q any, S any] struct {
	ApiRef types_app.API
	LogRef *zerolog.Logger

	Res          *MockResponseWriter
	Req          *http.Request
	LastError    error
	ResponseData *S
	RequestData  *Q
}

func NewMockApiHandler[Q any, S any](cfg *types_app.CLIConfig, mh *MockApiHandler[Q, S]) *MockApiHandler[Q, S] {
	api := NewMockApi(cfg)
	log := api.Log().With().Str("MockApiHandler", "MockApiHandler").Str("requestId", mh.RequestId())
	if mh.Req != nil && mh.Req.URL != nil {
		log.Str("requestPath", mh.Req.URL.Path)
	}

	mh.ApiRef = api
	logRef := log.Logger()
	mh.LogRef = &logRef

	mh.Res = NewMockResponseWriter()
	return mh
}

// type MockApiHandlerUntyped MockApiHandler[struct{}, struct{}]

func (m *MockApiHandler[Q, S]) Log() *zerolog.Logger {
	return m.LogRef
}

func (m *MockApiHandler[Q, S]) Api() types_app.API {
	return m.ApiRef
}
func (api *MockApiHandler[Q, S]) RequestId() string {
	return "test"
}
func (api *MockApiHandler[Q, S]) Response() http.ResponseWriter {
	return api.Res
}
func (api *MockApiHandler[Q, S]) Request() *http.Request {
	return api.Req
}
func (api *MockApiHandler[Q, S]) ErrorMsg(err error) {
	api.LastError = err
}

func (api *MockApiHandler[Q, S]) WriteMsg(res *S) {
	api.Response().WriteHeader(200)
	api.ResponseData = res
}

func (api *MockApiHandler[Q, S]) RequestMsg() (*Q, error) {
	return api.RequestData, nil
}

File: ts/src/testutils/api_handler.ts
import { APIMsg, Api, CLIConfig } from "../types/app";
import { HttpRequest, HttpResponseWriter, } from "../types";
import { MockApi } from "./api";
import { MockApp } from "./app";
import { MockResponseWriter } from "./response_writer";
import { Logger } from "@adviser/cement";

export class MockApiHandler<Q, S> implements APIMsg<Q, S> {
  readonly _log: Logger;
  readonly _api: Api;
  readonly _res: MockResponseWriter;
  readonly _req: HttpRequest;
  LastError?: Error;
  ResponseData?: S;
  RequestData?: Q;

  constructor(cfg: CLIConfig, req: HttpRequest) {
    const app = new MockApp({ CliConfig: cfg });
    this._api = new MockApi({ App: app });
    this._log = this._api.Log().With().Module("MockApiHandler").Str("requestId", this.RequestId()).Logger();
    this._res = new MockResponseWriter();
    this._req = req;
  }
  Log(): Logger {
    return this._log;
  }

  Api(): Api {
    return this._api;
  }
  RequestId(): string {
    return "test";
  }
  Response(): HttpResponseWriter {
    return this._res;
  }
  Request(): HttpRequest {
    return this._req;
  }
  ErrorMsg(err: Error): Promise<number> {
    this.LastError = err;
    return Promise.resolve(4711);
  }

  WriteMsg(res: S): Promise<number> {
    this.Response().WriteHeader(200);
    this.ResponseData = res;
    return Promise.resolve(42);
  }

  RequestMsg(): Promise<Q> {
    this._log.Debug().Any("msg", this.RequestData).Msg("RequestMsg");
    if (!this.RequestData) {
      return Promise.reject(new Error("RequestData not set"));
    }
    return Promise.resolve(this.RequestData);
  }
}
